containers/org.eclipse.linuxtools.docker.core/src/org/eclipse/linuxtools/internal/docker/core/DockerConnection.java

```diff
@@ -2196,44 +2196,26 @@ public void attachContainerOutput(final Closeable token, final String id,
 		}
 	}
 
-	@SuppressWarnings("unused")
 	public List<ContainerFileProxy> readContainerDirectory(final String id,
 			final String path) throws DockerException {
 		List<ContainerFileProxy> childList = new ArrayList<>();
+		String dirListing = null;
 		try {
 			DockerClient copyClient = getClientCopy();
 			final ExecCreation execCreation = copyClient.execCreate(id,
-					new String[] { "/bin/sh", "-c", "ls -l -F -Q " + path }, //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+					new String[] { "/bin/sh", "-c", //$NON-NLS-1$ //$NON-NLS-2$
+							"ls -l -F -Q " + path + "/" }, //$NON-NLS-1$
 					ExecCreateParam.attachStdout(),
 					ExecCreateParam.attachStderr());
 			final String execId = execCreation.id();
 			final LogStream pty_stream = copyClient.execStart(execId);
-			String lastLine = ""; //$NON-NLS-1$
+
 			try {
-				while (pty_stream.hasNext()) {
-					ByteBuffer b = pty_stream.next().content();
-					byte[] buffer = new byte[b.remaining()];
-					b.get(buffer);
-					String s = lastLine + new String(buffer);
-					String[] lines = s.split("\\r?\\n"); //$NON-NLS-1$
-					if (lines.length > 0) {
-						lastLine = lines[lines.length - 1];
-					} else {
-						lastLine = ""; //$NON-NLS-1$
-					}
-					for (int i = 0; i < lines.length - 1; ++i) {
-						String line = lines[i];
-						processDirectoryLine(line, path, childList);
-					}
-				}
-				if (!lastLine.isEmpty()) {
-					processDirectoryLine(lastLine, path, childList);
-				}
+				dirListing = pty_stream.readFully();
```

Comment:
The implementation of `readFully()` uses `hasNext()` - so I wasn't too worried:
```java
  public String readFully() {
    final StringBuilder stringBuilder = new StringBuilder();
    while (hasNext()) {
      stringBuilder.append(UTF_8.decode(next().content()));
    }
    return stringBuilder.toString();
  }
```
If there still is a problem it's a timing/concurrency issue. And in that case a different solution is needed (IMO).

Reply:
I believe it was a case of timing - the data was large and didn't all come back at once.  At the time there wasn't any way to fix it other than processing the lines and allowing more data to come along prior to the next hasNext() call.  Anyway, be aware as this will have to use the old method (even if temporarily) if the same problem still exists with the newer docker-client and dependencies as we will not regress functionality.  You can test by touching a 100000 or 1000000 empty files and copying the whole directory over.

Reply:
I did some experiments, even with lots of data (`/include` can become quite big) and never had an issue. Maybe it was a docker-issue that is fixed now.